relay-aws-extension/src/aws_extension.rs

```diff
@@ -0,0 +1,193 @@
+#[allow(dead_code)]
+struct Tracing {
+    pub r#type: String,
+    pub value: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[allow(dead_code)]
+#[serde(rename_all = "UPPERCASE", tag = "eventType")]
+enum NextEventResponse {
+    #[serde(rename_all = "camelCase")]
+    Invoke {
+        deadline_ms: u64,
+        request_id: String,
+        invoked_function_arn: String,
+        tracing: Tracing,
+    },
+    #[serde(rename_all = "camelCase")]
+    Shutdown {
+        shutdown_reason: String,
+        deadline_ms: u64,
+    },
+}
+
+#[derive(Debug, Fail)]
+#[fail(display = "aws extension error")]
+/// Generic error in an AWS extension context
+pub struct AwsExtensionError;
+
+/// Actor implementing an AWS extension
+///
+/// Spawns an actor that
+/// * registers with the AWS extensions API
+/// * sends blocking NextEvent calls to the extensions API to get the next invocation
+///
+/// Each finished invocation immediately polls for the next event. Note that AWS might
+/// freeze the container indefinitely for unused lambdas and this request will also wait
+/// till things are active again.
+pub struct AwsExtension {
+    /// The base url for the AWS Extensions API
+    base_url: String,
+    /// The reqwest client needs to be blocking and with 0 timeout
+    /// because the container might get frozen if lambda is unused
+    reqwest_client: Client,
+    /// The extension id that will be retrieved on register and used
+    /// for subsequent requests
+    extension_id: Option<String>,
+}
+
+impl AwsExtension {
+    /// Creates a new `AwsExtension` instance.
+    pub fn new(aws_runtime_api: String) -> Result<Self, AwsExtensionError> {
+        let base_url = format!("http://{}/2020-01-01/extension", aws_runtime_api);
+
+        let reqwest_client = reqwest::blocking::Client::builder()
+            .timeout(None)
+            .build()
+            .map_err(|_| AwsExtensionError)?;
+
+        Ok(AwsExtension {
+            base_url,
+            reqwest_client,
+            extension_id: None,
+        })
+    }
+
+    fn register(&mut self) -> Result<(), AwsExtensionError> {
+        let url = format!("{}/register", self.base_url);
+        let map = HashMap::from([("events", ["INVOKE", "SHUTDOWN"])]);
+
+        let res = self
+            .reqwest_client
+            .post(&url)
+            .header(EXTENSION_NAME_HEADER, EXTENSION_NAME)
+            .json(&map)
+            .send()
+            .map_err(|_| AwsExtensionError)?;
+
+        if res.status() != StatusCode::OK {
+            return Err(AwsExtensionError);
+        }
+
+        // let register_response: RegisterResponse = res.json().unwrap();
+
+        let extension_id = res
+            .headers()
+            .get(EXTENSION_ID_HEADER)
+            .and_then(|h| h.to_str().ok())
+            .map(|h| h.to_string())
+            .ok_or(AwsExtensionError)?;
+
+        self.extension_id = Some(extension_id);
+        relay_log::info!("AWS extension successfully registered");
```

Comment:
Question to relay folks: can we prefix all output of relay with a string like "[sentry-lambda-extension]"? This would help determining what output came from the extension and what came from the lambda function itself. (because both outputs end up in the same cloudwatch log stream)